### ISSUE TITLE
Bug: no-unnecessary-condition: false positives on zod branded string key for Record ??=

### ISSUE THREAD

**Issue by osdiab** [oxc-project/tsgolint#789]
repro: enable type aware linting with oxc. install zod. Do the following:

```
const brandedString = z.string().brand("brandedString")
const lookup: Record<z.infer<typeof brandedString>, number> = {};
lookup[brandedString.parse("test")] ??= 2;
```

result: Unnecessary optional chain on a non-nullish value.oxc[typescript-eslint(no-unnecessary-condition)](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-unnecessary-condition.html))

expected: no error, since no guarantee that this branded string would not be undefined. typescript seems to understand this.
